nixos/modules/services/home-automation/openthread-border-router.nix

```
@@ -0,0 +1,223 @@
+    };
+
+    logLevel = lib.mkOption {
+      type = lib.types.int;
+      default = 3;
+      description = ''
+        The level to use when logging messages. The levels are defined as follows:
+          EMERG 0
+          ALERT 1
+          CRIT 2
+          ERR 3
+          WARNING 4
+          NOTICE 5
+          INFO 6
+          DEBUG 7
+      '';
+    };
+
+    rest = {
+      listenAddress = lib.mkOption {
+        type = lib.types.str;
+        default = "::";
+        description = "The address on which to listen for REST API requests";
+        example = "::";
+      };
+
+      listenPort = lib.mkOption {
+        type = lib.types.port;
+        default = 8081;
+        description = "The port on which to listen for REST API requests";
+      };
+    };
+
+    web = {
+      enable = lib.mkEnableOption "enable the web interface";
+      listenAddress = lib.mkOption {
+        type = lib.types.str;
+        default = "::";
+        description = "The address on which the web interface should listen";
+        example = "::";
+      };
+
+      listenPort = lib.mkOption {
+        type = lib.types.port;
+        default = 8082;
+        description = "The port on which the web interface should listen";
+      };
+    };
+
+    radio = {
+      device = lib.mkOption {
+        type = lib.types.str;
+        default = "/dev/ttyUSB0";
+        description = "The device name of the serial port of the radio device. Ignored if url is manually set.";
+      };
+
+      baudRate = lib.mkOption {
+        type = lib.types.int;
+        default = 115200;
+        description = "The baud rate of the radio device. Ignored if url is manually set.";
+      };
+
+      flowControl = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = "Enable hardware flow control. Ignored if url is manually set.";
+      };
+
+      urlQueryString = lib.mkOption {
+        type = lib.types.str;
+        default = "";
+        description = "Extra URL query string parameters";
+        example = "bus-latency=100&region=ca";
+      };
+
+      url = lib.mkOption {
+        type = lib.types.str;
+        description = "The URL of the radio device to use";
+        example = "spinel+hdlc+uart:///dev/ttyUSB0?uart-baudrate=460800&uart-flow-control";
+      };
+
+      extraDevices = lib.mkOption {
+        type = lib.types.listOf lib.types.str;
+        default = [ ];
+        description = "Extra devices to add to the radio device";
+        example = "[ \"trel://eth0\" ]";
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.openthread-border-router.radio.url =
+      "spinel+hdlc+uart://${cfg.radio.device}?"
+      + lib.concatStringsSep "&" (
+        [ "uart-baudrate=${toString cfg.radio.baudRate}" ]
+        ++ lib.optional cfg.radio.flowControl "uart-flow-control"
+        ++ lib.optional (cfg.radio.urlQueryString != "") cfg.radio.urlQueryString
+      );
```

```suggestion
    services.openthread-border-router.radio.url = lib.mkDefault
      "spinel+hdlc+uart://${cfg.radio.device}?"
      + lib.concatStringsSep "&" (
        [ "uart-baudrate=${toString cfg.radio.baudRate}" ]
        ++ lib.optional cfg.radio.flowControl "uart-flow-control"
        ++ lib.optional (cfg.radio.urlQueryString != "") cfg.radio.urlQueryString
      );
```

Currently this will cause a conflict if `services.openthread-border-router.radio.url` is set elsewhere

---

Oh good catch. Do you know if there's a place to write eval-level tests for nixos modules?